File: evaluation/py/store.py
from __future__ import annotations

from dataclasses import dataclass
from threading import Lock


@dataclass
class Task:
    id: int
    title: str
    done: bool = False
    priority: str = "Medium"


class Store:
    def __init__(self):
        self._mu = Lock()
        self._tasks: list[Task] = []
        self._next_id = 1

    def list(self) -> list[Task]:
        with self._mu:
            return list(self._tasks)

    def list_by_priority(self, priority: str) -> list[Task]:
        with self._mu:
            return [t for t in self._tasks if t.priority == priority]

    def create(self, title: str, priority: str = "Medium") -> Task:
        with self._mu:
            t = Task(id=self._next_id, title=title, priority=priority or "Medium")
            self._next_id += 1
            self._tasks.append(t)
            return t

    def delete(self, id: int) -> bool:
        with self._mu:
            for i, t in enumerate(self._tasks):
                if t.id == id:
                    self._tasks.pop(i)
                    return True
            return False

    def toggle_done(self, id: int, done: bool) -> Task | None:
        with self._mu:
            for t in self._tasks:
                if t.id == id:
                    t.done = done
                    return t
            return None


VALID_PRIORITIES = {"Low", "Medium", "High"}


def is_valid_priority(p: str) -> bool:
    return p in VALID_PRIORITIES

File: evaluation/go/internal/store/store.go
package store

import "sync"

type Task struct {
	ID    int    `json:"id"`
	Title string `json:"title"`
	Done  bool   `json:"done"`
}

type Store struct {
	mu     sync.Mutex
	tasks  []Task
	nextID int
}

func New() *Store {
	return &Store{tasks: []Task{}, nextID: 1}
}

func (s *Store) List() []Task {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]Task, len(s.tasks))
	copy(out, s.tasks)
	return out
}

func (s *Store) Create(title string) Task {
	s.mu.Lock()
	defer s.mu.Unlock()
	t := Task{ID: s.nextID, Title: title, Done: false}
	s.nextID++
	s.tasks = append(s.tasks, t)
	return t
}

func (s *Store) Delete(id int) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	for i, t := range s.tasks {
		if t.ID == id {
			s.tasks = append(s.tasks[:i], s.tasks[i+1:]...)
			return true
		}
	}
	return false
}

func (s *Store) GetByID(id int) (Task, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, t := range s.tasks {
		if t.ID == id {
			return t, true
		}
	}
	return Task{}, false
}

func (s *Store) Update(id int, title *string, done *bool) (Task, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for i, t := range s.tasks {
		if t.ID == id {
			if title != nil {
				s.tasks[i].Title = *title
			}
			if done != nil {
				s.tasks[i].Done = *done
			}
			return s.tasks[i], true
		}
	}
	return Task{}, false
}
